src/array/cpu/spmat_op_impl_coo.cc

```diff
@@ -312,26 +313,65 @@ CSRMatrix COOToCSR(COOMatrix coo) {
   NDArray ret_indices;
   NDArray ret_data;
 
-  bool row_sorted = coo.row_sorted;
-  bool col_sorted = coo.col_sorted;
+  const bool row_sorted = coo.row_sorted;
+  const bool col_sorted = coo.col_sorted;
 
   if (row_sorted) {
     // compute indptr
-    IdType* Bp = static_cast<IdType*>(ret_indptr->data);
+    IdType* const Bp = static_cast<IdType*>(ret_indptr->data);
     Bp[0] = 0;
-    int64_t j = 0;
-    for (int64_t i = 0; i < N; ++i) {
-      const int64_t k = j;
-      for (; j < NNZ && row_data[j] == i; ++j) {}
-      Bp[i + 1] = Bp[i] + j - k;
+
+    if (!data) {
+      // Leave empty, and populate from inside of parallel block
+      coo.data = NDArray::Empty({NNZ}, coo.row->dtype, coo.row->ctx);
     }
+    IdType * const fill_data = data ? nullptr : static_cast<IdType*>(coo.data->data);
 
-    // TODO(minjie): Many of our current implementation assumes that CSR must have
-    //   a data array. This is a temporary workaround. Remove this after:
-    //   - The old immutable graph implementation is deprecated.
-    //   - The old binary reduce kernel is deprecated.
-    if (!COOHasData(coo))
-      coo.data = aten::Range(0, NNZ, coo.row->dtype.bits, coo.row->ctx);
+    if (NNZ > 0) {
+      #pragma omp parallel
+      {
+        const int num_threads = omp_get_num_threads();
+        const int thread_id = omp_get_thread_num();
+
+        const int64_t nz_chunk = (NNZ+num_threads-1)/num_threads;
+        const int64_t nz_start = thread_id*nz_chunk;
+        const int64_t nz_end = std::min(NNZ, nz_start+nz_chunk);
+
+        // each thread searchs the row array for a change, and marks it's
+        // location in Bp. Threads have overlapping ranges from nz_start-1 to
```

Comment:
Could you please elaborate more on the overlapping ranges?

Reply:
I expanded it and added an example in d824573 

```
341         // Each thread searches the row array for a change, and marks it's
342         // location in Bp. Threads, other than the first, start at the last
343         // index covered by the previous, in order to detect changes in the row
344         // array between thread partitions. This means that each thread after
345         // the first, searches the range [nz_start-1, nz_end). That is,
346         // if we had 10 non-zeros, and 4 threads, the indexes searched by each
347         // thread would be:
348         // 0: [0, 1, 2]
349         // 1: [2, 3, 4, 5]
350         // 2: [5, 6, 7, 8]
351         // 3: [8, 9]
352         //
353         // That way, if the row array were [0, 0, 1, 2, 2, 2, 4, 5, 5, 6], each
354         // change in row would be captured by one thread:
355         //
356         // 0: [0, 0, 1] - row 0
357         // 1: [1, 2, 2, 2] - row 1
358         // 2: [2, 4, 5, 5] - rows 2, 3, and 4
359         // 3: [5, 6] - rows 5 and 6
360         //
```